### DISCUSSION TITLE
Why use mysql container?

### DISCUSSION THREAD

**Discussion by nandoreisbr** [BretFisher/ama#193] [Q&A] [answered]
Hello Bret,
I'm starting to use docker in my application and it's going well with web servers.
I'm not sure in how or why to migrate my physical db hosts to docker.
Could you  please, talk about the advantages and disadvantages of using databases with docker?
Thank you

**Comment by BretFisher** [comment, marked as answer]
Lots of advantages, only a few disadvantages:

### Disadvantages
* If you're not careful about controlling resources then you'll starve your db or other apps on a single docker host. Docker has all the options to control that, but many forget to use them. This isn't a Docker-specific issue, but more of a "running many things on one host" issue.
* Similar to above, just because you can run 10 database servers on the same host with docker easily, doesn't mean you *should* do that, based on your resource requirements of each.
* If you need very high networking performance, you may not want to use Docker network plugins like overlay or bridge since they add overhead. If this becomes an issue, move to using `--network host`.
* You need to know docker well before you would want to run anything in it. Docker doesn't make running databases easier if you don't know the in's-and-out's of the docker tools to manage it.

### Advantages
* Makes it much easier and more reliable to update and re-deploy or move your database engines.
* Lets you use volumes and volume plugins to control the database persistent storage. This provides a nice abstraction layer where it's clear what is the db app, and what is the important storage you need to protect.
* Lets you easily spin up db's as part of a app stack using yaml files. This gets you closer to having dev setups look like prod setups.
* Gives you the option to, rather than many small db servers (one for each app), to choose to have a few big servers with many db containers on them. You can also still run just one db container (and nothing else) on a single node, if you so choose.
* Increased security. Just by running anything in a docker container rather than the host OS, you're using a restricted environment that might prevent a future security vulnerability from being an issue. Docker containers by default will use an apparmor profile, a seccomp profile, and more to make the app running in it less likely to have more rights then it needs.

And many more you can find on the web.

**Comment by nandoreisbr** [comment]
Thank you very much for you so completed answer. Now I'm sure that I'm going to deploy my database in container, and I'm sure too, that I have to learn more about docker, before to do that.

**Comment by sosensible** [comment]
When you spin up your db as a cluster DB, you cannot access the same `volume`, right? I did not see any comments about that in your pros and cons list. Do you have any thoughts about that?

**Comment by BretFisher** [comment]
Docker doesn't change the relationship of apps and their data files. A database container needs a volume to store its data, and those database files are unique to that running instance of a database binary.

Two database binaries can't use the same set of database files. Docker doesn't change that limitation. It's simply running an isolated binary and providing it storage for persistent data (volumes).

Different docker volume drivers provide different storage features around how a container accesses the volume (local, network shared, iscsi, etc.).

Does that answer your question?
